Add tools/ script to limit circleci to a set of jobs

driazati opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58001 Add tools/ script to limit circleci to a set of jobs**

Adds a script so that devs can generate a commit (at the base of a stack) that removes all CI jobs but the set that they care about. See CONTRIBUTING.md changes for usage

Differential Revision: [D28359910](https://our.internmc.facebook.com/intern/diff/D28359910)

driazati committed:
[skip ci] Add tools/ script to limit circleci to a set of jobs

[ghstack-poisoned]

driazati committed:
Update on "[skip ci] Add tools/ script to limit circleci to a set of jobs"




[ghstack-poisoned]

driazati committed:
Update on "[skip ci] Add tools/ script to limit circleci to a set of jobs"




[ghstack-poisoned]

driazati committed:
Update on "[skip ci] Add tools/ script to limit circleci to a set of jobs"


This adds a script that will edit `.circleci/config.yml` to only run a specific job (or jobs). Often there will be a specific CI failure that a developer wishes to debug before submitting the rest of their PR. Especially on tricker environments like Windows, if someone ends up "developing-in-CI" for to get results for one job, they end up incurring a ton of re-runs for other jobs.

The script would be to create the base PR in a stack on top of which the actual changes go:

```
python tools/circleci_only_job.py --job binary_linux_manywheel_3_6m_cpu_devtoolset7_nightly_test

git add .circleci/config.yml
git c -m"[skip ci][do not merge] edit .circleci/config.yml to only include 'my_cool_job'"

# make the actual changes
git c -m"My diff"

ghstack submit
```

The results of the script can be seen in #58002 which came from running `python tools/circleci_only_job.py --job binary_linux_manywheel_3_6m_cpu_devtoolset7_nightly_test`

Once done working on the top diff, the `[do not merge]` commit can be removed via an interactive rebase and the stack `ghstack submit`ed again.

This doesn't work for anything GitHub actions related and some jobs that don't follow the typical format (e.g. the doc builds/pushes) are just skipped so those also don't work. For GHA that could maybe be a follow up PR

[ghstack-poisoned]

driazati committed:
Update on "[skip ci] Add tools/ script to limit circleci to a set of jobs"


This adds a script that will edit `.circleci/config.yml` to only run a specific job (or jobs). Often there will be a specific CI failure that a developer wishes to debug before submitting the rest of their PR. Especially on tricker environments like Windows, if someone ends up "developing-in-CI" for to get results for one job, they end up incurring a ton of re-runs for other jobs.

The script would be to create the base PR in a stack on top of which the actual changes go:

```
python tools/circleci_only_job.py --job binary_linux_manywheel_3_6m_cpu_devtoolset7_nightly_test

git add .circleci/config.yml
git c -m"[skip ci][do not merge] edit .circleci/config.yml to only include 'my_cool_job'"

# make the actual changes
git c -m"My diff"

ghstack submit
```

The results of the script can be seen in #58002 which came from running `python tools/circleci_only_job.py --job binary_linux_manywheel_3_6m_cpu_devtoolset7_nightly_test`

Once done working on the top diff, the `[do not merge]` commit can be removed via an interactive rebase and the stack `ghstack submit`ed again.

This doesn't work for anything GitHub actions related and some jobs that don't follow the typical format (e.g. the doc builds/pushes) are just skipped so those also don't work. For GHA that could maybe be a follow up PR

[ghstack-poisoned]

driazati committed:
Update on "[skip ci] Add tools/ script to limit circleci to a set of jobs"


This adds a script that will edit `.circleci/config.yml` to only run a specific job (or jobs). Often there will be a specific CI failure that a developer wishes to debug before submitting the rest of their PR. Especially on tricker environments like Windows, if someone ends up "developing-in-CI" for to get results for one job, they end up incurring a ton of re-runs for other jobs.

The script would be to create the base PR in a stack on top of which the actual changes go:

```
python tools/explicit_ci_jobs.py --job binary_linux_manywheel_3_6m_cpu_devtoolset7_nightly_test --make-commit

# make the actual changes
git c -m"My diff"

ghstack submit
```

The results of the script can be seen in #58002 which came from running `python tools/circleci_only_job.py --job binary_linux_manywheel_3_6m_cpu_devtoolset7_nightly_test`

Once done working on the top diff, the `[do not merge]` commit can be removed via an interactive rebase and the stack `ghstack submit`ed again.

This doesn't work for anything GitHub actions related and some jobs that don't follow the typical format (e.g. the doc builds/pushes) are just skipped so those also don't work. For GHA that could maybe be a follow up PR

[ghstack-poisoned]

walterddr reviewed (commented):
qq: instead of making the specific-job change at the beginning (right after branch out from base), should we consider making this as a "cherry-pickable" commit that goes to the top of the stack? This way if I want to remove/change my CI preference I only need to alter the head commit.

drawback of this: making actual change to the diff can be less intuitive since you need to rollback a commit then amend/add your change

samestep commented:
This sounds really useful, thanks! The main thing I'm a bit concerned about is that people might accidentally forget to `git rebase -i` and land the whole stack (including the `[do not merge]` commit); is that possible? I can't remember whether `[do not merge]` actually prevents the PR from being merged.

samestep commented:
@walterddr Perhaps I'm misunderstanding, but wouldn't that cause CI to be re-run on _both_ commits in the stack, on every change? Wouldn't that defeat the purpose of this PR, which is to try to reduce the number of irrelevant re-runs while developing PRs?

samestep commented on tools/explicit_ci_jobs.py:
```diff
@@ -0,0 +1,134 @@
+import yaml
```
Could you add a shebang and also make this file executable?
```suggestion
#!/usr/bin/env python3
import yaml
```

samestep commented on tools/explicit_ci_jobs.py:
```diff
@@ -0,0 +1,134 @@
+    return new_workflows
+
+
+def commit_ci(files: List[str], message: str):
```
My local `mypy` is complaining about this line:
```
$ mypy --config=mypy-strict.ini tools/explicit_ci_jobs.py
tools/explicit_ci_jobs.py:84:1: error: Function is missing a return type annotation  [no-untyped-def]
Found 1 error in 1 file (checked 1 source file)
```
I'm guessing that's just because this PR's title contains `[skip ci]`?

samestep commented:
Another question I had (might be off-topic for this PR thread though, sorry): as we add more expensive GHA workflows to replace our existing CircleCI workflows, should we restrict them by `paths` so that they don't run on PRs that don't use them? Similar to this: https://github.com/pytorch/test-infra/pull/8#discussion_r624088930 @seemethere what are your thoughts?

samestep commented on tools/explicit_ci_jobs.py:
```diff
@@ -0,0 +1,134 @@
+
+    requires = job.get("requires", None)
+    if requires is not None:
+        for reuqirement in requires:
+            dependency = past_jobs[reuqirement]
```
```suggestion
        for requirement in requires:
            dependency = past_jobs[requirement]
```

samestep reviewed (approved):
lgtm other than the few comments I left, but most of those (other than the `mypy` one) aren't super important :)

driazati committed:
Update on "[skip ci] Add tools/ script to limit circleci to a set of jobs"


This adds a script that will edit `.circleci/config.yml` to only run a specific job (or jobs). Often there will be a specific CI failure that a developer wishes to debug before submitting the rest of their PR. Especially on tricker environments like Windows, if someone ends up "developing-in-CI" for to get results for one job, they end up incurring a ton of re-runs for other jobs.

The script would be to create the base PR in a stack on top of which the actual changes go:

```
python tools/explicit_ci_jobs.py --job binary_linux_manywheel_3_6m_cpu_devtoolset7_nightly_test --make-commit

# make the actual changes
git c -m"My diff"

ghstack submit
```

The results of the script can be seen in #58008 which came from running `python tools/circleci_only_job.py --job binary_linux_manywheel_3_6m_cpu_devtoolset7_nightly_test`. There are no tests on that PR due to the auto-added `[skip ci]`. The "real" changes are in #58009 which is where the actual change would be.

Once done working on the top diff, the `[do not merge]` commit can be removed via an interactive rebase and the stack `ghstack submit`ed again to run the full suite of tests.

This doesn't work for anything GitHub actions related and some jobs that don't follow the typical format (e.g. the doc builds/pushes) are just skipped so those also don't work. For GHA that could maybe be a follow up PR

[ghstack-poisoned]

driazati committed:
Update on "Add tools/ script to limit circleci to a set of jobs"




[ghstack-poisoned]

driazati committed:
Update on "Add tools/ script to limit circleci to a set of jobs"


Adds a script so that devs can generate a commit (at the base of a stack) that removes all CI jobs but the set that they care about. See CONTRIBUTING.md changes for usage 

[ghstack-poisoned]

driazati commented:
At that point you're basically emulating ghstack though right? And `branch_a` would still have the normal `config.yml` and `.github/workflows/*` so it would run the full gamut of builds/tests as far as I can tell

Either way a ghstack workflow FAQ would be good to do so we at least have something to point to
